ray_on_golem/server/services/golem/golem.py

```diff
@@ -196,30 +196,48 @@ async def _verify_ssh_connection(
         ssh_proxy_command: str,
         ssh_user: str,
         ssh_private_key_path: Path,
+        num_retries = 3,
+        retry_interval = 1,
     ) -> None:
-        ssh_command = get_ssh_command(ip, ssh_proxy_command, ssh_user, ssh_private_key_path)
+        ssh_command = f"{get_ssh_command(ip, ssh_proxy_command, ssh_user, ssh_private_key_path)} uptime"
 
-        logger.debug(f"Verifying ssh connection with:\n{ssh_command} uptime")
+        logger.debug(f"SSH connection check started on {activity}: cmd={ssh_command}")
 
-        process = await asyncio.create_subprocess_shell(
-            f"{ssh_command} uptime",
-            stdout=asyncio.subprocess.PIPE,
-            stderr=asyncio.subprocess.PIPE,
-        )
+        debug_data = ""
+
+        async def check():
+            nonlocal debug_data
+
+            process = await asyncio.create_subprocess_shell(
+                ssh_command,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+
+            stdout, stderr = await process.communicate()
 
-        stdout, stderr = await process.communicate()
+            debug_data = f"{activity=}, exitcode={process.returncode}, {stdout=}, {stderr=}"
 
-        msg = f"Validation of ssh connection done on {activity}:\n[cmd exited with {process.returncode}]"
-        if stdout:
-            msg += f" [stdout] {stdout.decode().strip()}"
-        if stderr:
-            msg += f" [stderr] {stderr.decode().strip()}"
+            if process.returncode != 0:
+                raise Exception(f"SSH connection check failed. {debug_data}")
 
-        if process.returncode != 0:
-            logger.warning(msg)
-            raise Exception(msg)
+        retry = num_retries
 
-        logger.info(msg)
+        while retry > 0:
+            try:
+                await check()
+                break
+            except Exception as e:
+                retry -= 1
+                if retry:
+                    logger.warning(f"{str(e)}, retrying {retry}...")
```

Comment:
This can be `logger.debug` we were already considering making all failed ssh of `DEBUG` level

Reply:
hmm... why? it does sound "warning"-like ... I mean, it _should_ work normally... and maybe it _is_ a sign of something amiss?

Reply:
Because there are many of them already and with retries there will be more.

Reply:
but it's still abnormal... imvho...